### PR TITLE
Wp/release/0.7.9

### PR DESCRIPTION
**Pull request by vyshakhbabji**

### CODE REVIEW

File: src/qos.ts [hunk 1]
```diff
@@ -79,7 +79,7 @@ const publishQosStats = (session: WebPhoneSession, qosStatsObj: QosStats, option
     const event = options.event || 'vq-rtcpxr';
     options.expires = 60;
     options.contentType = 'application/vq-rtcpxr';
-    options.extraHeaders = (options.extraHeaders || []).concat(session.ua.defaultHeaders);
+    options.extraHeaders = options.extraHeaders || [];
```
[the remark this answers is not in the document]
Comment (vyshakhbabji) [reply src/qos.ts]: Fixed

File: src/session.ts [hunk 2]
```diff
@@ -545,18 +591,13 @@ async function warmTransfer(
...
-
     transferOptions.extraHeaders = (transferOptions.extraHeaders || []).concat(this.ua.defaultHeaders);
-
     return this.refer(target, transferOptions);
 }
 
 /*--------------------------------------------------------------------------------------------------------------------*/
 
-async function transfer(this: WebPhoneSession, target: WebPhoneSession, options: any = {}): Promise<ReferClientContext> {
-    options.extraHeaders = (options.extraHeaders || []).concat(this.ua.defaultHeaders);
```
[the remark this answers is not in the document]
Comment (vyshakhbabji) [reply src/session.ts]: Fixed.

File: src/session.ts [hunk 3]
```diff
@@ -520,15 +521,60 @@ function dtmf(this: WebPhoneSession, dtmf: string, duration = 100, interToneGap
...
+        return;
+    }
+    var options = {
+        modifiers: [],
+        extraHeaders: [],
+        sessionDescriptionHandlerOptions: undefined
+    };
+
+    if (this.pendingReinvite) {
+        this.logger.warn('Hold in progress. Please wait until complete, then try again.');
```
[the remark this answers is not in the document]
Comment (vyshakhbabji) [reply src/session.ts]: Fixed

File: src/session.ts [hunk 4]
```diff
@@ -520,15 +521,60 @@ function dtmf(this: WebPhoneSession, dtmf: string, duration = 100, interToneGap
...
+        return;
+    }
+    if (!this.sessionDescriptionHandler) {
+        this.logger.warn("No SessionDescriptionHandler, can't put the call on hold.");
```
[the remark this answers is not in the document]
Comment (vyshakhbabji) [reply src/session.ts]: Fixed

File: src/session.ts [hunk 5]
```diff
@@ -520,15 +521,60 @@ function dtmf(this: WebPhoneSession, dtmf: string, duration = 100, interToneGap
...
+        return;
+    }
+
+    if (this.sessionDescriptionHandler) {
```
[the remark this answers is not in the document]
Comment (vyshakhbabji) [reply src/session.ts]: Fixed

File: src/session.ts [hunk 6]
```diff
@@ -520,15 +521,60 @@ function dtmf(this: WebPhoneSession, dtmf: string, duration = 100, interToneGap
...
+            return Promise.reject(e);
+        });
 }
 
 /*--------------------------------------------------------------------------------------------------------------------*/
 
-function unhold(this: WebPhoneSession): Promise<any> {
-    return setLocalHold(this, false);
+async function unhold(this: WebPhoneSession): Promise<any> {
+    await this.__unhold();
```
[the remark this answers is not in the document]
Comment (vyshakhbabji) [reply src/session.ts]: Fixed

**Comment by vyshakhbabji** [comment]
Release Notes:
1. Better handling of  Hold / Unhold feature 
2. Full support for Warm Transfer 
3. BugFix for Blind Transfer
4. Improvements for Re-invites
5. Added additional logging

File: src/session.ts [hunk 7]
```diff
@@ -543,19 +642,25 @@ async function warmTransfer(
...
-    transferOptions.extraHeaders = (transferOptions.extraHeaders || []).concat(this.ua.defaultHeaders);
-
-    return this.refer(target, transferOptions);
+    return new Promise(resolve => {
+        transferOptions.extraHeaders = (transferOptions.extraHeaders || []).concat(this.ua.defaultHeaders);
+        transferOptions.receiveResponse = response => {
+            if (response.statusCode === 202) resolve(response);
+        };
+        this.logger.log('Completing warm transfer');
+        this.refer(target, transferOptions);
```
Comment (kirill-konshin) [inline src/session.ts]: Same here

File: src/session.ts [hunk 8]
```diff
@@ -733,7 +838,7 @@ function addTrack(this: WebPhoneSession, remoteAudioEle, localAudioEle): void {
             } else if (!isNoAudio(report)) {
                 this.noAudioReportCount = 0;
             }
-        }, 2000)
+        }, 2000);
```
Comment (kirill-konshin) [inline src/session.ts]: What this magic number means here?

Comment (vyshakhbabji) [reply src/session.ts]: This is to detect no audio for 6 seconds -> every 2 sec * 3 counts the code checks for mediastats  before initiating reinvite.

Comment (kirill-konshin) [reply src/session.ts]: You need to name this constant then

File: src/session.ts (hunk 7 above)
Comment (vyshakhbabji) [reply src/session.ts]: I was trying to resolve when I get response of sip 202 for refer. If this is not necessary I will just keep this as is and return ( Promise.resolve(  this.refer(target, transferOptions)) )

File: src/session.ts [hunk 9]
```diff
@@ -520,21 +522,118 @@ function dtmf(this: WebPhoneSession, dtmf: string, duration = 100, interToneGap
...
-function blindTransfer(this: WebPhoneSession, target, options = {}): Promise<ReferClientContext> {
-    return Promise.resolve(this.refer(target, options));
+async function blindTransfer(this: WebPhoneSession, target, options = {}): Promise<ReferClientContext> {
+    return new Promise(resolve => {
+        // @ts-ignore
+        options.receiveResponse = response => {
+            if (response.statusCode === 202) resolve(response);
+        };
+        this.logger.log('Call transfer initiated');
+        this.refer(target, options);
```
[the remark this answers is not in the document]
Comment (vyshakhbabji) [reply src/session.ts]: I was trying to resolve when I get response of sip 202 for refer. If this is not necessary I will just keep this as is and return ( Promise.resolve( this.refer(target, transferOptions)) )

File: src/session.ts [hunk 10]
```diff
@@ -520,21 +522,118 @@ function dtmf(this: WebPhoneSession, dtmf: string, duration = 100, interToneGap
...
+                        return _this.receiveReinviteResponse(response);
+                    }
+                });
+            })
+            .catch(e => {
+                _this.pendingReinvite = false;
+                _this.logger.log('Unhold not completed');
+                reject(e);
+            });
+    });
```
[the remark this answers is not in the document]
Comment (vyshakhbabji) [reply src/session.ts]: Please verify. Made changes necessary.

File: src/session.ts [hunk 11]
```diff
@@ -520,21 +522,118 @@ function dtmf(this: WebPhoneSession, dtmf: string, duration = 100, interToneGap
...
+        if (!this.sessionDescriptionHandler) {
+            throw new Error("No SessionDescriptionHandler, can't put the call on hold.");
+        }
+
+        options.modifiers.push(this.sessionDescriptionHandler.holdModifier);
+        this.pendingReinvite = true;
+
+        this.sessionDescriptionHandler
+            .getDescription(options.sessionDescriptionHandlerOptions, options.modifiers)
+            .then(description => {
```
[the remark this answers is not in the document]
Comment (vyshakhbabji) [reply src/session.ts]: Please verify. Made changes necessary.

File: src/session.ts [hunk 12]
```diff
@@ -520,21 +522,118 @@ function dtmf(this: WebPhoneSession, dtmf: string, duration = 100, interToneGap
...
+                _this.sendRequest(C.INVITE, {
```
[the remark this answers is not in the document]
Comment (vyshakhbabji) [reply src/session.ts]: Please verify. Made changes necessary.

File: src/session.ts [hunk 13]
```diff
@@ -520,21 +522,118 @@ function dtmf(this: WebPhoneSession, dtmf: string, duration = 100, interToneGap
...
-}
+async function hold(this: WebPhoneSession): Promise<any> {
+    return new Promise((resolve, reject) => {
+        var _this = this;
+        if (this.status !== Session.C.STATUS_WAITING_FOR_ACK && this.status !== Session.C.STATUS_CONFIRMED) {
+            throw new Exceptions.InvalidStateError(this.status);
+        }
+        if (this.localHold) {
+            this.logger.warn('Session already on hold');
+            return Promise.resolve();
```
[the remark this answers is not in the document]
Comment (vyshakhbabji) [reply src/session.ts]: Done

File: src/session.ts [hunk 14]
```diff
@@ -520,21 +522,118 @@ function dtmf(this: WebPhoneSession, dtmf: string, duration = 100, interToneGap
 
 /*--------------------------------------------------------------------------------------------------------------------*/
 
-function hold(this: WebPhoneSession): Promise<any> {
-    this.stopMediaStats();
-    return setLocalHold(this, true);
-}
+async function hold(this: WebPhoneSession): Promise<any> {
+    return new Promise((resolve, reject) => {
```
[the remark this answers is not in the document]
Comment (vyshakhbabji) [reply src/session.ts]: Please verify. Made changes necessary.

File: src/session.ts [hunk 15]
```diff
@@ -520,15 +521,60 @@ function dtmf(this: WebPhoneSession, dtmf: string, duration = 100, interToneGap
...
+
+    if (this.sessionDescriptionHandler) {
+        options.modifiers.push(this.sessionDescriptionHandler.holdModifier);
+    }
+
+    this.pendingReinvite = true;
+
+    this.sessionDescriptionHandler
+        .getDescription(options.sessionDescriptionHandlerOptions, options.modifiers)
+        .then(function (description) {
```
[the remark this answers is not in the document]
Comment (vyshakhbabji) [reply src/session.ts]: done

File: src/session.ts (hunk 8 above)
Comment (vyshakhbabji) [reply src/session.ts]: Added

File: src/session.ts [hunk 16]
```diff
@@ -520,20 +527,87 @@ function dtmf(this: WebPhoneSession, dtmf: string, duration = 100, interToneGap
...
 
 /*--------------------------------------------------------------------------------------------------------------------*/
 
-function unhold(this: WebPhoneSession): Promise<any> {
-    return setLocalHold(this, false);
+async function unhold(this: WebPhoneSession): Promise<any> {
+    if (this.status !== Session.C.STATUS_WAITING_FOR_ACK && this.status !== Session.C.STATUS_CONFIRMED) {
+        throw new Exceptions.InvalidStateError(this.status);
+    }
+    if (!this.localHold) {
```
Comment (kirill-konshin) [inline src/session.ts]: Session can also be on hold when hold is initiated by another party. I don't think it's safe to rely on local var.

File: src/session.ts [hunk 17]
```diff
@@ -520,20 +527,87 @@ function dtmf(this: WebPhoneSession, dtmf: string, duration = 100, interToneGap
 
 /*--------------------------------------------------------------------------------------------------------------------*/
 
-function hold(this: WebPhoneSession): Promise<any> {
+async function sendReinvite(this: WebPhoneSession, options: any = {}): Promise<any> {
+    if (options === void 0) {
```
Comment (kirill-konshin) [inline src/session.ts]: Options will be defaulted already in function args definition

File: src/session.ts [hunk 18]
```diff
@@ -664,7 +739,7 @@ function onLocalHold(this: WebPhoneSession): boolean {
 
 function addTrack(this: WebPhoneSession, remoteAudioEle, localAudioEle): void {
     const pc = this.sessionDescriptionHandler.peerConnection;
-
+    const mediaCheckTimer = 2000;
```
Comment (kirill-konshin) [inline src/session.ts]: Constants should be defined outside functions, at the top of the file. Otherwise you initialize it every time without a need.

Comment (vyshakhbabji) [reply src/session.ts]: Oh yeah correct! will move it out.

File: src/session.ts (hunk 17 above)
Comment (vyshakhbabji) [reply src/session.ts]: Sure will change this.

File: src/session.ts (hunk 16 above)
Comment (vyshakhbabji) [reply src/session.ts]: This is going to be local for the session. I am pretty sure ,  the session put on hold by other party  will not affects localhold - i can reconfirm .

**Pull request merged**